wagtail/admin/rich_text/converters/editor_html.py

```diff
@@ -0,0 +1,168 @@
+
+
+# Whitelist rules which are always active regardless of the rich text features that are enabled
+
+BASE_WHITELIST_RULES = {
+    '[document]': allow_without_attributes,
+    'p': allow_without_attributes,
+    'div': allow_without_attributes,
+    'br': allow_without_attributes,
+}
+
+
+class DbWhitelister(Whitelister):
+    """
+    A custom whitelisting engine to convert the HTML as returned by the rich text editor
+    into the pseudo-HTML format stored in the database (in which images, documents and other
+    linked objects are identified by ID rather than URL):
+
+    * implements a 'construct_whitelister_element_rules' hook so that other apps can modify
+      the whitelist ruleset (e.g. to permit additional HTML elements beyond those in the base
+      Whitelister module);
+    * replaces any element with a 'data-embedtype' attribute with an <embed> element, with
+      attributes supplied by the handler for that type as defined in embed_handlers;
+    * rewrites the attributes of any <a> element with a 'data-linktype' attribute, as
+      determined by the handler for that type defined in link_handlers, while keeping the
+      element content intact.
+    """
+    def __init__(self, converter_rules):
+        self.converter_rules = converter_rules
+        self.element_rules = BASE_WHITELIST_RULES.copy()
+        for rule in self.converter_rules:
+            if isinstance(rule, WhitelistRule):
+                self.element_rules[rule.element] = rule.handler
+
+        # apply legacy construct_whitelister_element_rules hooks to the assembled list
+        construct_whitelist_hooks = hooks.get_hooks('construct_whitelister_element_rules')
+        if construct_whitelist_hooks:
+            warnings.warn(
+                'The construct_whitelister_element_rules hook is deprecated and will be removed '
+                'in Wagtail 2.2. Use register_rich_text_features instead',
+                RemovedInWagtail22Warning
+            )
+
+            for fn in construct_whitelist_hooks:
+                self.element_rules.update(fn())
+
+    @cached_property
+    def embed_handlers(self):
+        embed_handlers = {}
+        for rule in self.converter_rules:
+            if isinstance(rule, EmbedTypeRule):
+                embed_handlers[rule.embed_type] = rule.handler
+
+        return embed_handlers
+
+    @cached_property
+    def link_handlers(self):
+        link_handlers = {}
+        for rule in self.converter_rules:
+            if isinstance(rule, LinkTypeRule):
+                link_handlers[rule.link_type] = rule.handler
+
+        return link_handlers
+
+    def clean_tag_node(self, doc, tag):
+        if 'data-embedtype' in tag.attrs:
+            embed_type = tag['data-embedtype']
+            # fetch the appropriate embed handler for this embedtype
+            try:
+                embed_handler = self.embed_handlers[embed_type]
+            except KeyError:
+                # discard embeds with unrecognised embedtypes
```

Comment:
I'm undecided about how severe we want to make this error... I suspect we don't want it to fail _loudly_ (because it would affect the site front-end too), and a log message / warning seems too easily ignored to be worth adding. Given that uncertainty, I'd prefer to keep the existing behaviour (silently failing).